refactor AND bugfix

wraithgar opened the pull request:
Two commits here, one is a refactor so that we could more easily find the bug, the second is the bugfix.

When installing a relative path into a workspace, arborist was trying to install it relative to the project root.  This was ok before workspaces as that was the only place you could add new deps.  With workspaces we need to now add new deps relative to the tree it's being added to.  The second commit does this.


```sh
~/D/n/s/ws $ npm pkg delete dependencies.a -w b
{}
~/D/n/s/ws $ node ../../cli i ./workspaces/a -w b

up to date, audited 5 packages in 871ms

found 0 vulnerabilities
~/D/n/s/ws $ npm pkg get dependencies -w b
{
  "b": {
    "a": "file:../a"
  }
}
~/D/n/s/ws $ npm ls
ws@1.0.0 /Users/wraithgar/Development/npm/scratch/ws
├── a@1.0.0 -> ./workspaces/a
└─┬ b@1.0.0 -> ./workspaces/b
  └── a@1.0.0 deduped -> ./workspaces/a
```

wraithgar committed:
fix: inline single-use functions

wraithgar commented:
This was refactoring and cleanup trying to debug an issue, the bug still stands but this code allowed us to find where the bug starts, to be fixed later.

ljharb commented on workspaces/arborist/lib/add-rm-pkg-deps.js:
```diff
@@ -4,8 +4,67 @@ const log = require('proc-log')
-  for (const spec of add) {
-    addSingle({ pkg, spec, saveBundle, saveType })
+  for (const { name, rawSpec } of add) {
+    // if the user does not give us a type, we infer which type(s)
+    // to keep based on the same order of priority we do when
+    // building the tree as defined in the _loadDeps method of
+    // the node class.
+    if (!saveType) {
+      saveType = inferSaveType(pkg, name)
+    }
```
instead of this, maybe `saveType = inferSaveType(pkg, name)` in the signature? unless you expect a user to give you a non-undefined falsy value that you want to preserve (like an empty string)

wraithgar replied:
that's outside the scope of this PR, it was just inlining existing code

ljharb commented on workspaces/arborist/lib/add-rm-pkg-deps.js:
```diff
@@ -103,9 +97,8 @@ const inferSaveType = (pkg, name) => {
-const { hasOwnProperty } = Object.prototype
 const hasSubKey = (pkg, depType, name) => {
-  return pkg[depType] && hasOwnProperty.call(pkg[depType], name)
+  return pkg[depType] && Object.prototype.hasOwnProperty.call(pkg[depType], name)
```
this was actually an important thing - to cache hasOwnProperty at module level. Whenever arborist is used as a library, you can't assume Object.prototype is unmodified.

wraithgar replied:
This is the same logic that made people use `void 0` instead of `undefined` this is not the kind of stuff we should have to worry about.  If you have things in your tree rewriting `Object.prototype` that's ... a different problem.

wraithgar replied:
`lib/arborist/reify.js` already calls `Object.prototype` so it's not even consistent here.  If this was something we cared about it'd be a linting rule.

ljharb replied:
It's the same logic, and *both* are things that libraries absolutely have to worry about. If your code breaks because something did something dumb elsewhere, it's still broken.

This isn't something that's trivial to lint against.

wraithgar committed:
fix: create links relative to the target

Added link deps need to be relative to the package they're being added
to, not the project root.  In the past the project root was the only
place you could add things but workspaces changed this.
